sansa-inference-spark/src/main/scala/net/sansa_stack/inference/spark/forwardchaining/axioms/ForwardRuleReasonerOWLHorst.scala

```diff
@@ -0,0 +1,650 @@
+package net.sansa_stack.inference.spark.forwardchaining.axioms
+
+import java.io.File
+
+import net.sansa_stack.inference.utils.CollectionUtils
+import net.sansa_stack.owl.spark.rdd.{FunctionalSyntaxOWLAxiomsRDDBuilder, OWLAxiomsRDD}
+import org.apache.spark.SparkContext
+import org.apache.spark.rdd.{RDD, UnionRDD}
+import org.apache.spark.sql.SparkSession
+import org.semanticweb.owlapi.apibinding.OWLManager
+import org.semanticweb.owlapi.model._
+
+import scala.collection.JavaConverters._
+
+
+class ForwardRuleReasonerOWLHorst (sc: SparkContext, parallelism: Int = 2) extends TransitiveReasoner{
+
+  def this(sc: SparkContext) = this(sc, sc.defaultParallelism)
+
+  def apply(axioms: RDD[OWLAxiom], input: String): RDD[OWLAxiom] = {
+
+    val startTime = System.currentTimeMillis()
+
+    val manager = OWLManager.createOWLOntologyManager()
+    val dataFactory = manager.getOWLDataFactory
+
+    val axiomsRDD = axioms.cache() // cache this RDD because it will be used quite often
+
+    // ------------ extract the schema elements -------------------
+    val classes: RDD[OWLClass] = axiomsRDD.flatMap {
+      case axiom: HasClassesInSignature => axiom.classesInSignature().iterator().asScala
+      case _ => null
+    }.filter(_ != null).distinct()
+
+    var subClassof = extractAxiom(axiomsRDD, AxiomType.SUBCLASS_OF)
+    var subDataProperty = extractAxiom(axiomsRDD, AxiomType.SUB_DATA_PROPERTY)
+    var subObjProperty = extractAxiom(axiomsRDD, AxiomType.SUB_OBJECT_PROPERTY)
+    var subAnnProp = extractAxiom(axiomsRDD, AxiomType.SUB_ANNOTATION_PROPERTY_OF)
+    val objectProDomain = extractAxiom(axiomsRDD, AxiomType.OBJECT_PROPERTY_DOMAIN)
+    val objectProRange = extractAxiom(axiomsRDD, AxiomType.OBJECT_PROPERTY_RANGE)
+    val dataProDomain = extractAxiom(axiomsRDD, AxiomType.DATA_PROPERTY_DOMAIN)
+    val dataProRange = extractAxiom(axiomsRDD, AxiomType.DATA_PROPERTY_RANGE)
+    val equClass = extractAxiom(axiomsRDD, AxiomType.EQUIVALENT_CLASSES)
+    var equDataProp = extractAxiom(axiomsRDD, AxiomType.EQUIVALENT_DATA_PROPERTIES)
+    val equObjProp = extractAxiom(axiomsRDD, AxiomType.EQUIVALENT_OBJECT_PROPERTIES)
+
+    // --------- extract instance elements
+    var classAssertion = extractAxiom(axiomsRDD, AxiomType.CLASS_ASSERTION)
+    var dataPropAssertion = extractAxiom(axiomsRDD, AxiomType.DATA_PROPERTY_ASSERTION).asInstanceOf[RDD[OWLDataPropertyAssertionAxiom]]
+    var objPropAssertion = extractAxiom(axiomsRDD, AxiomType.OBJECT_PROPERTY_ASSERTION).asInstanceOf[RDD[OWLObjectPropertyAssertionAxiom]]
+
+    // ---------------- Schema Rules --------------------
+    // 1. we have to process owl:equivalentClass (resp. owl:equivalentProperty) before computing the transitive closure
+    // of rdfs:subClassOf (resp. rdfs:sobPropertyOf)
+    // O11a: (C owl:equivalentClass D) -> (C rdfs:subClassOf D )
```

Comment:
Could you please document somewhere what rule set you are using? For example I didn't find any rule 11a in the standard OWL Horst paper [Combining RDF and Part of OWL with Rules: Semantics, Decidability, Complexity](https://link.springer.com/content/pdf/10.1007/11574620_48.pdf). Later you switch to completely different rule identifiers (`R1`, `R2`, ...). I'd suggest to stick to ter Horst's rule identifiers if possible.

Reply:
Then it would have been good if this was noted somewhere in the code. I added the comments.